### ISSUE TITLE
Review bsf for destructor consistancy

### ISSUE THREAD

**Issue by jonesmz** [GameFoundry/bsf#69]
A good rule of thumb is that a class should be or have one of the following

- final
- protected non-virtual destructor
- public virtual destructor

This is especially important for classes that have member variables and derive from a base class. They run the risk of being deleted with the destructors for their member variables not being called if the base class is not marked as virtual.

A first example would be DynLib.h, which has a public non-virtual destructor but isn't final. A class could derive from DynLib with a member variable, and then not have it's member variable destructed upon delete.

**Comment by BearishSun** [comment]
Makes sense, I agree. As a rule of thumb, everything without a virtual destructor isn't generally intended to be overridden and can probably be marked with final. Except when it is overridden, in which case it's likely some struct without any actual virtual methods in which case the destructor should be made private.
